### PR TITLE
[TSan] Unlocking MonoStats, GCStats and JITGCStats

### PR DESCRIPTION
**Pull request by cherusker**
As the `Unlocked* ()` functions have been approved in https://github.com/mono/mono/pull/5310, I used them to blacklist some of the most obvious counting variables whose data races should be ignored: `MonoStats`, `GCStats` and `JITGCStats` (more to follow if approved).

Please find details about all changes within the commit messages.

### CODE REVIEW

File: mono/metadata/sgen-mono.c [hunk 1]
```diff
@@ -2799,18 +2800,19 @@ mono_gc_add_memory_pressure (gint64 value)
  */
 
 void
-sgen_client_degraded_allocation (size_t size)
+sgen_client_degraded_allocation (void)
```
Comment (lambdageek) [inline mono/metadata/sgen-mono.c]: @BrzVlad is the `size` arg of `sgen_client_degraded_allocation` not useful?

Comment (alexrp) [inline mono/metadata/sgen-mono.c]: Anything starting with `sgen_client_` is part of the SGen API, so this is a breaking change.

File: mono/utils/unlocked.h [hunk 2]
```diff
@@ -39,4 +39,32 @@ UnlockedIncrementSize (gsize *val)
 	return ++*val;
 }
 
+MONO_UNLOCKED_ATTRS
```
Comment (alexrp) [inline mono/utils/unlocked.h]: Please split the changes to this file into a separate commit.

Comment (cherusker) [inline mono/utils/unlocked.h]: @alexrp sure; done!

File: mono/metadata/sgen-mono.c (hunk 1 above)
Comment (BrzVlad) [inline mono/metadata/sgen-mono.c]: @alexrp I don't think these are meant to be an official API and they were changed in the past. Are there other runtimes out there that actually use sgen in the first place ?

Comment (alexrp) [inline mono/metadata/sgen-mono.c]: I guess @schani would know. He did the whole SGen separation effort for a use case I can't remember.

File: mono/sgen/sgen-gc.c [hunk 3]
```diff
@@ -3053,9 +3055,7 @@ sgen_gc_collect (int generation)
 int
 sgen_gc_collection_count (int generation)
 {
-	if (generation == 0)
-		return gc_stats.minor_gc_count;
-	return gc_stats.major_gc_count;
+	return UnlockedRead (generation == 0 ? &gc_stats.minor_gc_count : &gc_stats.major_gc_count);
```
Comment (cherusker) [inline mono/sgen/sgen-gc.c]: @luhenry can we compare with `== GENERATION_NURSERY` here?

File: mono/metadata/sgen-mono.c (hunk 1 above)
Comment (schani) [inline mono/metadata/sgen-mono.c]: Changing this is fine.

**Comment by cherusker** [comment]
@vargaz the `Unlocked* ()` functions have absolutely no performance impact (if compiled without the TSan setup); if, however, the affected JIT functions are not as performance critical as the rest - should we / can we use the [`Interlocked* ()`](https://github.com/mono/mono/blob/master/mono/utils/atomic.h) functions there instead of the `Unlocked* ()` ones?

**Comment by vargaz** [comment]
The non-perf critical ones can use the simple Interlocked functions.

File: mono/metadata/boehm-gc.c [hunk 4]
```diff
@@ -458,7 +459,7 @@ on_gc_notification (GC_EventType event)
 		if (mono_perfcounters)
 			mono_perfcounters->gc_collections0++;
 #endif
-		gc_stats.major_gc_count ++;
+		UnlockedIncrement (&gc_stats.major_gc_count);
```
Comment (lewurm) [inline mono/metadata/boehm-gc.c]: can be `Interlocked`

File: mono/metadata/boehm-gc.c [hunk 5]
```diff
@@ -482,7 +483,7 @@ on_gc_notification (GC_EventType event)
 			mono_perfcounters->gc_gen0size = heap_size;
 		}
 #endif
-		gc_stats.major_gc_time += mono_100ns_ticks () - gc_start_time;
+		UnlockedAddUnsigned64 (&gc_stats.major_gc_time, mono_100ns_ticks () - gc_start_time);
```
Comment (lewurm) [inline mono/metadata/boehm-gc.c]: can be `Interlocked`

File: mono/metadata/loader.c [hunk 6]
```diff
@@ -1658,7 +1659,7 @@ mono_get_method_from_token (MonoImage *image, guint32 token, MonoClass *klass,
 		methods_size += sizeof (MonoMethod);
 	}
 
-	mono_stats.method_count ++;
+	UnlockedIncrementSize (&mono_stats.method_count);
```
Comment (lewurm) [inline mono/metadata/loader.c]: can be `Interlocked`

File: mono/metadata/sgen-mono.c [hunk 7]
```diff
@@ -2799,18 +2800,19 @@ mono_gc_add_memory_pressure (gint64 value)
...
-	static int num_degraded = 0;
+	static gint32 last_major_gc_warned = -1;
+	static gint32 num_degraded = 0;
 
-	if (last_major_gc_warned < (int)gc_stats.major_gc_count) {
-		++num_degraded;
-		if (num_degraded == 1 || num_degraded == 3)
+	gint32 major_gc_count = UnlockedRead (&gc_stats.major_gc_count);
+	if (UnlockedRead (&last_major_gc_warned) < major_gc_count) {
+		gint32 num = UnlockedIncrement (&num_degraded);
```
Comment (lewurm) [inline mono/metadata/sgen-mono.c]: everything can use `Interlocked` here

File: mono/mini/driver.c [hunk 8]
```diff
@@ -1420,7 +1421,7 @@ mono_jit_parse_options (int argc, char * argv[])
 			opt->break_on_exc = TRUE;
 		} else if (strcmp (argv [i], "--stats") == 0) {
 			mono_counters_enable (-1);
-			mono_stats.enabled = TRUE;
+			UnlockedWriteBool (&mono_stats.enabled, TRUE);
```
Comment (lewurm) [inline mono/mini/driver.c]: can be `Interlocked`

File: mono/mini/driver.c [hunk 9]
```diff
@@ -1767,7 +1768,7 @@ mono_main (int argc, char* argv[])
 			mono_print_vtable = TRUE;
 		} else if (strcmp (argv [i], "--stats") == 0) {
 			mono_counters_enable (-1);
-			mono_stats.enabled = TRUE;
+			UnlockedWriteBool (&mono_stats.enabled, TRUE);
```
Comment (lewurm) [inline mono/mini/driver.c]: can be `Interlocked`

File: mono/mini/mini.c [hunk 10]
```diff
@@ -3125,9 +3126,9 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
 
 	if (opts & MONO_OPT_GSHARED) {
 		if (try_generic_shared)
-			mono_stats.generics_sharable_methods++;
+			UnlockedIncrementSize (&mono_stats.generics_sharable_methods);
```
Comment (lewurm) [inline mono/mini/mini.c]: can be `Interlocked`

File: mono/mini/mini.c [hunk 11]
```diff
@@ -3125,9 +3126,9 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
...
 		else if (mono_method_is_generic_impl (method))
-			mono_stats.generics_unsharable_methods++;
+			UnlockedIncrementSize (&mono_stats.generics_unsharable_methods);
```
Comment (lewurm) [inline mono/mini/mini.c]: can be `Interlocked`

File: mono/mini/mini.c [hunk 12]
```diff
@@ -4236,9 +4237,9 @@ mono_jit_compile_method_inner (MonoMethod *method, MonoDomain *target_domain, in
 		code = cfg->native_code;
 
 		if (cfg->gshared && mono_method_is_generic_sharable (method, FALSE))
-			mono_stats.generics_shared_methods++;
+			UnlockedIncrementSize (&mono_stats.generics_shared_methods);
```
Comment (lewurm) [inline mono/mini/mini.c]: can be `Interlocked`

File: mono/mini/mini.c [hunk 13]
```diff
@@ -4236,9 +4237,9 @@ mono_jit_compile_method_inner (MonoMethod *method, MonoDomain *target_domain, in
...
 		if (cfg->gsharedvt)
-			mono_stats.gsharedvt_methods++;
+			UnlockedIncrementSize (&mono_stats.gsharedvt_methods);
```
Comment (lewurm) [inline mono/mini/mini.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 14]
```diff
@@ -1701,7 +1703,7 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
 	TV_GETTIME (last_minor_collection_start_tv);
 	atv = last_minor_collection_start_tv;
 
-	binary_protocol_collection_begin (gc_stats.minor_gc_count, GENERATION_NURSERY);
+	binary_protocol_collection_begin (UnlockedRead (&gc_stats.minor_gc_count), GENERATION_NURSERY);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 15]
```diff
@@ -1729,7 +1731,7 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
 	degraded_mode = 0;
 	objects_pinned = 0;
 
-	SGEN_LOG (1, "Start nursery collection %d %p-%p, size: %d", gc_stats.minor_gc_count, nursery_section->data, nursery_section->end_data, (int)(nursery_section->end_data - nursery_section->data));
+	SGEN_LOG (1, "Start nursery collection %" G_GINT32_FORMAT " %p-%p, size: %d", UnlockedRead (&gc_stats.minor_gc_count), nursery_section->data, nursery_section->end_data, (int)(nursery_section->end_data - nursery_section->data));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 16]
```diff
@@ -1744,7 +1746,7 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
 	init_gray_queue (&gc_thread_gray_queue);
 	ctx = CONTEXT_FROM_OBJECT_OPERATIONS (object_ops_nopar, &gc_thread_gray_queue);
 
-	gc_stats.minor_gc_count ++;
+	UnlockedIncrement (&gc_stats.minor_gc_count);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 17]
```diff
@@ -1861,9 +1863,9 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
 	major_collector.finish_nursery_collection ();
 
 	TV_GETTIME (last_minor_collection_end_tv);
-	gc_stats.minor_gc_time += TV_ELAPSED (last_minor_collection_start_tv, last_minor_collection_end_tv);
+	UnlockedAddUnsigned64 (&gc_stats.minor_gc_time, TV_ELAPSED (last_minor_collection_start_tv, last_minor_collection_end_tv));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 18]
```diff
@@ -1861,9 +1863,9 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
...
 
-	sgen_debug_dump_heap ("minor", gc_stats.minor_gc_count - 1, NULL);
+	sgen_debug_dump_heap ("minor", UnlockedRead (&gc_stats.minor_gc_count) - 1, NULL);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 19]
```diff
@@ -1896,7 +1898,7 @@ collect_nursery (const char *reason, gboolean is_overflow, SgenGrayQueue *unpin_
 			time_minor_scan_los - los_scan_start,
 			time_minor_finish_gray_stack - finish_gray_start);
 
-	binary_protocol_collection_end (gc_stats.minor_gc_count - 1, GENERATION_NURSERY, 0, 0);
+	binary_protocol_collection_end (UnlockedRead (&gc_stats.minor_gc_count) - 1, GENERATION_NURSERY, 0, 0);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 20]
```diff
@@ -2147,7 +2149,7 @@ major_start_collection (SgenGrayQueue *gc_thread_gray_queue, const char *reason,
 {
 	SgenObjectOperations *object_ops_nopar, *object_ops_par = NULL;
 
-	binary_protocol_collection_begin (gc_stats.major_gc_count, GENERATION_OLD);
+	binary_protocol_collection_begin (UnlockedRead (&gc_stats.major_gc_count), GENERATION_OLD);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 21]
```diff
@@ -2178,8 +2180,8 @@ major_start_collection (SgenGrayQueue *gc_thread_gray_queue, const char *reason,
 	check_scan_starts ();
 
 	degraded_mode = 0;
-	SGEN_LOG (1, "Start major collection %d", gc_stats.major_gc_count);
-	gc_stats.major_gc_count ++;
+	SGEN_LOG (1, "Start major collection %" G_GINT32_FORMAT, UnlockedRead (&gc_stats.major_gc_count));
+	UnlockedIncrement (&gc_stats.major_gc_count);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: both can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 22]
```diff
@@ -2295,7 +2297,7 @@ major_finish_collection (SgenGrayQueue *gc_thread_gray_queue, const char *reason
 	TV_GETTIME (atv);
 	time_major_sweep += TV_ELAPSED (btv, atv);
 
-	sgen_debug_dump_heap ("major", gc_stats.major_gc_count - 1, reason);
+	sgen_debug_dump_heap ("major", UnlockedRead (&gc_stats.major_gc_count) - 1, reason);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 23]
```diff
@@ -2327,7 +2329,7 @@ major_finish_collection (SgenGrayQueue *gc_thread_gray_queue, const char *reason
                         time_major_scan_mod_union_los - los_scan_start,
                         time_major_finish_gray_stack - finish_gray_start);
 
-	binary_protocol_collection_end (gc_stats.major_gc_count - 1, GENERATION_OLD, counts.num_scanned_objects, counts.num_unique_scanned_objects);
+	binary_protocol_collection_end (UnlockedRead (&gc_stats.major_gc_count) - 1, GENERATION_OLD, counts.num_scanned_objects, counts.num_unique_scanned_objects);
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 24]
```diff
@@ -2355,7 +2357,7 @@ major_do_collection (const char *reason, gboolean is_overflow, gboolean forced)
 	sgen_gray_object_queue_dispose (&gc_thread_gray_queue);
 
 	TV_GETTIME (time_end);
-	gc_stats.major_gc_time += TV_ELAPSED (time_start, time_end);
+	UnlockedAddUnsigned64 (&gc_stats.major_gc_time, TV_ELAPSED (time_start, time_end));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 25]
```diff
@@ -2391,7 +2393,7 @@ major_start_concurrent_collection (const char *reason)
 	num_objects_marked = major_collector.get_and_reset_num_major_objects_marked ();
 
 	TV_GETTIME (time_end);
-	gc_stats.major_gc_time += TV_ELAPSED (time_start, time_end);
+	UnlockedAddUnsigned64 (&gc_stats.major_gc_time, TV_ELAPSED (time_start, time_end));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 26]
```diff
@@ -2419,7 +2421,7 @@ major_update_concurrent_collection (void)
 	sgen_los_update_cardtable_mod_union ();
 
 	TV_GETTIME (total_end);
-	gc_stats.major_gc_time += TV_ELAPSED (total_start, total_end);
+	UnlockedAddUnsigned64 (&gc_stats.major_gc_time, TV_ELAPSED (total_start, total_end));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 27]
```diff
@@ -2441,7 +2443,7 @@ major_finish_concurrent_collection (gboolean forced)
 	sgen_workers_stop_all_workers (GENERATION_OLD);
 
 	SGEN_TV_GETTIME (time_major_conc_collection_end);
-	gc_stats.major_gc_time_concurrent += SGEN_TV_ELAPSED (time_major_conc_collection_start, time_major_conc_collection_end);
+	UnlockedAddUnsigned64 (&gc_stats.major_gc_time_concurrent, SGEN_TV_ELAPSED (time_major_conc_collection_start, time_major_conc_collection_end));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c [hunk 28]
```diff
@@ -2456,7 +2458,7 @@ major_finish_concurrent_collection (gboolean forced)
 	sgen_gray_object_queue_dispose (&gc_thread_gray_queue);
 
 	TV_GETTIME (total_end);
-	gc_stats.major_gc_time += TV_ELAPSED (total_start, total_end);
+	UnlockedAddUnsigned64 (&gc_stats.major_gc_time, TV_ELAPSED (total_start, total_end));
```
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: can be `Interlocked`

File: mono/sgen/sgen-gc.c (hunk 3 above)
Comment (lewurm) [inline mono/sgen/sgen-gc.c]: yes you can.  also, you can use `Interlocked` here.

File: mono/metadata/boehm-gc.c (hunk 5 above)
Comment (cherusker) [inline mono/metadata/boehm-gc.c]: @vargaz @lewurm can I convert the `gc_stats.*_gc_time`_* values ([gc-internal-agnostic.h](https://github.com/mono/mono/pull/5381/commits/d87f71c42f3cd2ea0f4e0097952fe1f4b51585db#diff-91d6a369b7a39d55fd63be9665e86498)) to `gint64`? The reason behind this: An `InterlockedAdd64 ()` function exists already, however, I would have to "invent" `InterlockedAddUnsigned64 ()`. Furthermore, @vargaz told me at to simply use signed types if there is no explicit reason for using unsigned ones (https://github.com/mono/mono/pull/5299) and `9,223,372,036,854,775,807 ns` are still almost `300 years` ... do we really need `600 years`?

Comment (lewurm) [inline mono/metadata/boehm-gc.c]: should be fine.  also, `mono_100ns_ticks ()` returns a `gint64`.

File: mono/utils/atomic.h [hunk 29]
```diff
@@ -464,4 +464,13 @@ extern void InterlockedWritePointer(volatile gpointer *dst, gpointer val);
 #define InterlockedAddP(p,add) InterlockedAdd64 ((volatile gint64*)p, (gint64)add)
 #endif
 
+/* The following functions cannot be found on any platform, and thus they can be declared without further existence checks */
+
+static inline void
+InterlockedWriteBool (volatile gboolean *dest, gboolean val)
```
Comment (cherusker) [inline mono/utils/atomic.h]: @BrzVlad @alexrp is it OK to put this function here?

**Comment by cherusker** [comment]
The most recent update changed a lot of functions from `Unlocked* ()` to `Interlocked* ()` and also changed the types of `MonoStats` variables from `gsize` to `gint32`. The reason for that is explained in the commit message: mainly it's about `Interlocked* ()` being defined for `gint32` but not for `gsize`. If this is unwanted, however, I can either

* go back to `gsize` and add `Interlocked*Size ()` methods to `atomics.h` or
* we can go with `gint64`.

**Comment by vargaz** [comment]
I think we should use either gint32 or gint64, depending on the expected values.

**Comment by cherusker** [comment]
@vargaz I have seen these values going into the low millions but always staying far away from the billions with `class_vtable_size` usually leading. After all, I only reduced the capacity by the factor `2` (no `log` reduction or worse) - so generally it should not matter. Anyways, `gint64` would be safer of course if it's OK to use it for performance-critical code thinking about the overhead of 64-bit operations on 32-bit machines JUST for monitoring. @kumpera do you have any thoughts on this?

**Comment by vargaz** [comment]
I think gint64 is only needed for things like amount of GC memory allocated.

**Comment by lewurm** [comment]
@monojenkins squash

**Pull request merged**